From 21d3eb76b68d6758122eddaf00b5823f9b2d7f34 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Luk=C3=A1=C5=A1=20Tyrychtr?= <ltyrycht@redhat.com>
Date: Tue, 6 Oct 2026 15:29:33 +0200
Subject: [PATCH] a11y: Return error for getChildAtIndex on the root

Match the behavior of other AT SPI2 accessible interface implementations
(e.g. the general one) and return an invalid index count when an invalid
toplevel gets requested.

Fixes #8444.

Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/10440>
---
 gtk/a11y/gtkatspiroot.c | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/gtk/a11y/gtkatspiroot.c b/gtk/a11y/gtkatspiroot.c
index d8cb6fb0166..7ca2a288921 100644
--- a/gtk/a11y/gtkatspiroot.c
+++ b/gtk/a11y/gtkatspiroot.c
@@ -302,6 +302,7 @@ handle_accessible_method (GDBusConnection       *connection,
       g_variant_get (parameters, "(i)", &idx);
 
       GtkWidget *window = NULL;
+      GtkWidget *child = NULL;
       guint n_toplevels = g_list_model_get_n_items (self->toplevels);
       for (guint i = 0; i < n_toplevels; i++)
         {
@@ -313,15 +314,24 @@ handle_accessible_method (GDBusConnection       *connection,
             continue;
 
           if (real_idx == idx)
-            break;
+            {
+              child = window;
+              break;
+            }
 
           real_idx += 1;
         }
 
-      if (window == NULL)
-        return;
+      if (child == NULL)
+        {
+          g_dbus_method_invocation_return_error (invocation,
+                                                 G_IO_ERROR,
+                                                 G_IO_ERROR_INVALID_ARGUMENT,
+                                                 "No child with index %d", idx);
+          return;
+        }
 
-      GtkATContext *context = gtk_accessible_get_at_context (GTK_ACCESSIBLE (window));
+      GtkATContext *context = gtk_accessible_get_at_context (GTK_ACCESSIBLE (child));
 
       const char *name = g_dbus_connection_get_unique_name (self->connection);
       const char *path = gtk_at_spi_context_get_context_path (GTK_AT_SPI_CONTEXT (context));
